denlillemand/week3/ass4/app/mainlist.jsx

```diff
@@ -0,0 +1,113 @@
+import React from 'react';
+import NamedList from './namedlist.jsx';
+import NamedListForm from './namedlistform.jsx';
+
+export default class MainList extends React.Component {
+    constructor(props, context) {
+        super(props, context);
+        this.state = {};
+        this.swapMessage = this.swapMessage.bind(this);
+        this.archiveMessage = this.archiveMessage.bind(this);
+        this.onSubmitNamedList = this.onSubmitNamedList.bind(this);
+        this.deleteMessage = this.deleteMessage.bind(this);
+        this.onSubmitMessage = this.onSubmitMessage.bind(this);
+    }
+
+    archiveMessage(id, currentList, isArchived) {
+        var mutatedArray = this.state[currentList].map(function (message) {
+            if (message.id === id) {
+                var clonedMessage = Object.assign({}, message);
+                clonedMessage.archived = !isArchived;
+                return clonedMessage;
+            }
+
+            return message;
+        });
+
+        this.setState({
+            [currentList]: mutatedArray,
+        });
+    }
+
+    swapMessage(messageId, destinationList, currentList) {
+        let messageIndex = -1;
+        this.state[currentList].forEach(function (message, index, array) {
+            if (message.id === messageId) {
+                messageIndex = index;
+            }
+        });
+
+        if (messageIndex > -1) {
+            let copiedCurrentList = this.state[currentList].slice();
+            let removedMessage = copiedCurrentList.splice(messageIndex, 1);
+            if (removedMessage[0].archived) {
+                alert('Cannot remove an archived message');
+            } else {
+                let copiedDestinationList = this.state[destinationList].slice();
+                this.setState({
+                    [currentList]: copiedCurrentList,
+                    [destinationList]: copiedDestinationList.concat(removedMessage),
+                });
+            }
+        } else {
+            alert("Couldn't find the message id you wanted to remove");
+        }
+    };
+
+    onSubmitNamedList(name) {
+        if (this.state[name]) {
+            alert('Cannot create another list with the same name');
+        } else {
+            this.setState({
+                [name]: [],
```

Comment:
I think this is one of the issues with your data modeling choice. There is no clear boundary where the your `namedLists` live. In addition it seems strange to me that an empty `NamedList` is just an empty array.